docs/design/enhance-pluggable-policy.md

```diff
@@ -0,0 +1,175 @@
+  - user - the user-string which a user was authenticated as.  This is included in the Context.
+  - groups - the groups to which the user belongs.  This is included in the Context.
+  - verb - string describing the requestion action. Today we have: get, list, watch, create, update, and delete
+  - what kind of object is being accessed 
+    - applies only to the API endpoints, such as 
+        `/api/v1beta1/pods`.  For miscelaneous endpoints, like `/version`, the kind is the empty string.
+  - the namespace of the object being access, or the empty string if the endpoint does not support namespaced objects.  This is included in the Context.
+  - content - the runtime.Object being submitted. This only has content for creates and updates
+  - resourceName - the name of the resource being gotten (get-ed) or deleted.
+
+### Authorizer Interface
+The existing Authorizer interface is very simple, but there isn't a way to provide details about allows, denies, or failures.  The extended detail is useful for UIs that want to describe why certain actions are disallowed.  Not all Authorizers will want to provide that information, but for those that do having that capability is useful.  In addition, adding a `GetAllowedSubjects` method that returns back the users and groups that can perform a particular action makes it possible to answer questions like, "who can see resources in my namespace" (see [ResourceAccessReview](#ResourceAccessReview) further down).
+
+```
+// OLD
+type Authorizer interface {
+  Authorize(a Attributes) error 
+}
+```
+
+```
+// NEW
+type Authorizer interface {
+  // Authorize takes a Context (for namespace, user, and traceability) and Attributes to make a policy determination.
+  // reason is an optional return value that can describe why a policy decision was made.
+  Authorize(ctx api.Context, a Attributes) (allowed bool, reason string, evaluationError error)
+
+  // GetAllowedSubjects takes a Context (for namespace and traceability) and Attributes to determine which users and 
+  // groups are allowed to perform the described action in the namespace.    This API enables the ResourceBasedReview requests below
+  GetAllowedSubjects(ctx api.Context, a Attributes) (users util.StringSet, groups util.StringSet, evaluationError error)
+}
+```
+
+### SubjectAccessReview
+`/api/{version}/ns/{namespace}/subjectAccessReview` - This API answers the question: can a user or group (use authenticated user if none is specified) perform a given action.  Given the Authorizer interface (proposed or existing), this endpoint can be implemented generically against any Authorizer by creating the correct Attributes and making an .Authorize() call.
+
+SubjectAccessReview is runtime.Object with associated RESTStorage that only accepts creates.  The caller POSTs a SubjectAccessReview to this URL and he gets a SubjectAccessReviewResponse back.  Here is an example of a call and its corresponding return.
```

Comment:
We may want to make this a SubjectAccessResponse or unify the struct between request and response - action being one nested struct and response being the other (or just use status).